File: code/backend/src/llm_geometry/geo/tokenizer.py
"""Deterministic word-level tokenizer for the Geometry Lab.

Lowercases, normalizes typographic punctuation (curly quotes → ASCII), and splits
into word / number / single-punctuation tokens. The vocabulary is the top-1000 token
types of the training corpus by frequency (ties broken alphabetically, so building is
fully deterministic), preceded by the three specials ``<unk>=0 <eos>=1 <pad>=2``.
Punctuation marks count as "word types" here — they are among the most frequent
types and an LM over the corpus is meaningless without them.

``encode`` marks out-of-vocab positions (``unk``) and truncates at CONTEXT_WINDOW per
the frozen `/api/geo/tokenize` contract; ``truncate_side`` chooses which end survives
("right" keeps the first 50 tokens — the tokenize-preview behavior; "left" keeps the
last tokens — what LM conditioning wants). ``to_json``/``from_json`` round-trip the
exact vocabulary.
"""

from __future__ import annotations

import json
import re
from collections import Counter
from dataclasses import dataclass
from functools import lru_cache
from typing import Any

from ..errors import InvalidParamError
from .config import (
    CONTEXT_WINDOW,
    EOS_ID,
    EOS_TOKEN,
    PAD_ID,
    PAD_TOKEN,
    UNK_ID,
    UNK_TOKEN,
    VOCAB_SIZE,
    VOCAB_WORDS,
)
from .corpus import load_corpus_text

# Typographic → ASCII normalization applied before splitting.
_NORMALIZE = str.maketrans(
    {
        "’": "'",  # ’
        "‘": "'",  # ‘
        "“": '"',  # “
        "”": '"',  # ”
        " ": " ",  # nbsp
    }
)

# Words (with internal apostrophes: "alice's", "won't"), numbers, or any single
# non-space symbol (punctuation, dashes, underscores, …).
_TOKEN_RE = re.compile(r"[a-z]+(?:'[a-z]+)*|[0-9]+|[^\sa-z0-9]")

_SPECIAL_TEXTS = {UNK_ID: UNK_TOKEN, EOS_ID: EOS_TOKEN, PAD_ID: PAD_TOKEN}

#: The one vocabulary format a model file may carry (``GeoTokenizer.to_json``).
_VOCAB_FORMAT = "geo-tokenizer-v1"

#: Special-token ids a vocabulary may declare, under either spelling. Mirrors
#: ``geoEngine/tokenizer.fromModelVocabJson`` — see :meth:`GeoTokenizer.from_json`.
_EXPECTED_SPECIALS = {
    UNK_TOKEN: UNK_ID,
    EOS_TOKEN: EOS_ID,
    PAD_TOKEN: PAD_ID,
    "unk": UNK_ID,
    "eos": EOS_ID,
    "pad": PAD_ID,
}


def _validate_specials(specials: Any) -> None:
    """Refuse a vocabulary whose declared special ids are not the ones we use.

    Ignoring the block (which is what "parse only `words`" amounts to) let a file say
    ``"<unk>": 5`` and load with 200 while every id it labels is read with ``<unk> = 0``.
    """
    if specials is None:
        return
    if not isinstance(specials, dict):
        raise InvalidParamError(
            f"vocabulary `specials` must be an object, got {type(specials).__name__}"
        )
    for token, expected in _EXPECTED_SPECIALS.items():
        declared = specials.get(token)
        if declared is not None and declared != expected:
            raise InvalidParamError(
                f"vocabulary: special {token} has id {declared!r}, expected {expected}"
            )


# Tokens that should not be preceded by a space when detokenizing.
_NO_SPACE_BEFORE = set(".,;:!?)]}'…") | {"''"}
_NO_SPACE_AFTER = set("([{")


def split_words(text: str) -> list[str]:
    """Deterministically split ``text`` into lowercase word/punctuation tokens."""
    return _TOKEN_RE.findall(text.lower().translate(_NORMALIZE))


@dataclass(frozen=True)
class EncodedText:
    ids: list[int]
    texts: list[str]
    unk: list[bool]
    n_unk: int
    truncated: bool

    def tokens(self) -> list[dict]:
        """Contract-shaped token list: [{"id", "text", "unk"}, …]."""
        return [{"id": i, "text": t, "unk": u} for i, t, u in zip(self.ids, self.texts, self.unk)]


class GeoTokenizer:
    """Fixed-vocabulary word-level tokenizer (deterministic, JSON-serializable)."""

    def __init__(self, words: list[str]) -> None:
        if len(words) != VOCAB_WORDS:
            raise InvalidParamError(
                f"GeoTokenizer requires exactly {VOCAB_WORDS} words, got {len(words)}"
            )
        if len(set(words)) != len(words):
            raise InvalidParamError("GeoTokenizer vocabulary contains duplicates")
        self.words = list(words)
        self.id_to_text: dict[int, str] = dict(_SPECIAL_TEXTS)
        for offset, word in enumerate(self.words):
            self.id_to_text[len(_SPECIAL_TEXTS) + offset] = word
        self.text_to_id: dict[str, int] = {
            text: tid for tid, text in self.id_to_text.items() if tid not in _SPECIAL_TEXTS
        }
        assert len(self.id_to_text) == VOCAB_SIZE

    # -- construction ----------------------------------------------------------------

    @classmethod
    def from_corpus_text(cls, text: str) -> "GeoTokenizer":
        counts = Counter(split_words(text))
        # Deterministic order: by descending frequency, then alphabetically.
        ranked = sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))
        words = [w for w, _ in ranked[:VOCAB_WORDS]]
        if len(words) < VOCAB_WORDS:
            raise InvalidParamError(
                f"Corpus has only {len(words)} distinct token types; " f"{VOCAB_WORDS} are required"
            )
        return cls(words)

    # -- encode / decode -------------------------------------------------------------

    def encode(
        self,
        text: str,
        *,
        truncate: bool = True,
        truncate_side: str = "right",
        max_tokens: int = CONTEXT_WINDOW,
    ) -> EncodedText:
        pieces = split_words(text)
        truncated = False
        if truncate and len(pieces) > max_tokens:
            truncated = True
            pieces = pieces[:max_tokens] if truncate_side == "right" else pieces[-max_tokens:]
        ids: list[int] = []
        unk: list[bool] = []
        for piece in pieces:
            tid = self.text_to_id.get(piece)
            if tid is None:
                ids.append(UNK_ID)
                unk.append(True)
            else:
                ids.append(tid)
                unk.append(False)
        return EncodedText(ids=ids, texts=pieces, unk=unk, n_unk=sum(unk), truncated=truncated)

    def encode_stream(self, text: str) -> list[int]:
        """Encode without truncation (training/fine-tuning token streams)."""
        return self.encode(text, truncate=False).ids

    def decode(self, ids: list[int]) -> str:
        out: list[str] = []
        for tid in ids:
            text = self.id_to_text.get(int(tid))
            if text is None:
                raise InvalidParamError(f"Token id {tid} is out of range (0..{VOCAB_SIZE - 1})")
            if not out:
                out.append(text)
            elif text in _NO_SPACE_BEFORE or out[-1] in _NO_SPACE_AFTER:
                out.append(text)
            else:
                out.append(" " + text)
        return "".join(out)

    # -- serialization ---------------------------------------------------------------

    def to_json(self) -> str:
        """The CANONICAL vocabulary serialization — byte-identical in both builds.

        `vocab_sha256` is a digest of these exact bytes, so the spelling is part of the
        format, not a formatting preference. Python used ``", "``/``": "`` separators
        while the browser used ``JSON.stringify``'s compact ones with its own key
        order, and the same model therefore had two different digests depending on
        which build saved it. Pinned here (and in
        ``lib/geoEngine/tokenizer.canonicalVocabJson``) as: keys sorted, compact
        separators, ``ensure_ascii`` so the bytes do not depend on the encoding.
        """
        return json.dumps(
            {
                "format": "geo-tokenizer-v1",
                "specials": {UNK_TOKEN: UNK_ID, EOS_TOKEN: EOS_ID, PAD_TOKEN: PAD_ID},
                "words": self.words,
            },
            ensure_ascii=True,
            sort_keys=True,
            separators=(",", ":"),
        )

    @classmethod
    def from_json(cls, payload: str) -> "GeoTokenizer":
        """Parse the canonical serialization, refusing anything else with a TYPED error.

        This is the vocabulary half of loading a model file, and it used to be three
        unguarded lines: ``json.loads``, ``data.get``, ``list(data["words"])``. Seven
        malformed ``vocab`` blocks therefore reached the user as untyped HTTP 500s whose
        whole message was a Python exception string (``'list' object has no attribute
        'get'``, ``unhashable type: 'list'``, ``'words'``) — on a file the user chose, so
        the one thing the message had to say was which part of THEIR file was wrong.

        ``specials`` is validated rather than ignored, and the ``tokens``-shaped export is
        refused rather than crashing, because the browser engine
        (``geoEngine/tokenizer.fromModelVocabJson``) does exactly this: while Python
        ignored ``specials`` a file declaring ``<unk> = 5`` loaded here with 200 and was
        refused there, so the two stacks disagreed about whether a file was valid at all.
        """
        if not isinstance(payload, str):
            raise InvalidParamError(f"a vocabulary must be JSON text, got {type(payload).__name__}")
        try:
            data = json.loads(payload)
        except ValueError as exc:  # JSONDecodeError
            raise InvalidParamError(f"vocabulary is not valid JSON: {exc}")
        if not isinstance(data, dict):
            raise InvalidParamError(
                "vocabulary must be a JSON object with a `words` array, got "
                f"{type(data).__name__}"
            )
        if data.get("format") != _VOCAB_FORMAT:
            raise InvalidParamError(
                f"Unknown tokenizer format {data.get('format')!r} (expected " f"{_VOCAB_FORMAT!r})"
            )
        _validate_specials(data.get("specials"))
        words = data.get("words")
        if words is None and "tokens" in data:
            raise InvalidParamError(
                "vocabulary has `tokens` but no `words`: `tokens` is the static site's "
                "asset shape (specials included), while a model file carries the "
                f"{_VOCAB_FORMAT!r} `words` list of exactly {VOCAB_WORDS} entries"
            )
        if not isinstance(words, list) or any(not isinstance(w, str) for w in words):
            raise InvalidParamError("vocabulary `words` must be an array of strings")
        return cls(words)


@lru_cache(maxsize=1)
def get_tokenizer() -> GeoTokenizer:
    """The canonical tokenizer, built deterministically from the real corpus."""
    return GeoTokenizer.from_corpus_text(load_corpus_text())


def tokenizer_for(weights_token: str | None, store: Any = None) -> GeoTokenizer:
    """The tokenizer whose words give THIS model's token ids their meaning.

    Models trained from scratch on a user's own text carry their own vocabulary, so
    reading their ids with the canonical one would label every token wrongly — and so
    does anything DERIVED from such a model, because a fine-tune or a weight edit
    changes the numbers, not what the ids mean (`weights.inherited_vocab`). Only sets
    descended from the canonical checkpoint have no stored vocabulary, and those are
    the ones the canonical tokenizer is right for.

    A token this store does not have RAISES (``weights.weight_set_entry``); it does not
    fall back. The fallback was the same corruption reached through a different door: an
    evicted or unknown token answered 200 from ``GET /api/geo/tokenize`` with the user's
    ids read under Alice in Wonderland's word list, while ``GET /api/geo/trace`` answered
    404 for the identical request — so the two routes disagreed about whether the model
    existed, and the one that said yes was the one that was wrong.
    """
    if not weights_token or weights_token == "learned":
        return get_tokenizer()
    from .weights import load_weight_set_vocab

    payload = load_weight_set_vocab(weights_token, store=store)
    return GeoTokenizer.from_json(payload) if payload else get_tokenizer()

File: code/frontend/src/lib/geoEngine/tokenizer.ts
/**
 * Deterministic word-level tokenizer — exact TypeScript port of
 * code/backend/src/llm_geometry/geo/tokenizer.py.
 *
 * Lowercases, normalizes typographic punctuation (curly quotes -> ASCII, nbsp ->
 * space), then splits into word / number / single-punctuation tokens with the same
 * regex as the backend: `[a-z]+(?:'[a-z]+)*|[0-9]+|[^\sa-z0-9]`. The `u` flag makes
 * `[^\s...]` match by code point like Python (an astral character is one token, not
 * two surrogate halves). Known residual divergence (documented, not observed in the
 * corpus/goldens): JS `\s` also treats U+FEFF as whitespace while Python's does not.
 *
 * Vocabulary: 1000 corpus word types + specials <unk>=0 <eos>=1 <pad>=2, loaded from
 * the backend-exported vocab.json (GeoTokenizer.to_json()).
 */

import { invalidParam } from "./errors";
import {
  CONTEXT_WINDOW,
  EOS_ID,
  EOS_TOKEN,
  PAD_ID,
  PAD_TOKEN,
  UNK_ID,
  UNK_TOKEN,
  VOCAB_SIZE,
  VOCAB_WORDS,
} from "./model";

// Typographic -> ASCII normalization applied before splitting (backend _NORMALIZE).
const NORMALIZE_RE = /[’‘“” ]/g;
const NORMALIZE_MAP: Record<string, string> = {
  "’": "'",
  "‘": "'",
  "“": '"',
  "”": '"',
  " ": " ",
};

// Words (with internal apostrophes), numbers, or any single non-space symbol.
const TOKEN_RE = /[a-z]+(?:'[a-z]+)*|[0-9]+|[^\sa-z0-9]/gu;

const SPECIAL_TEXTS: ReadonlyMap<number, string> = new Map([
  [UNK_ID, UNK_TOKEN],
  [EOS_ID, EOS_TOKEN],
  [PAD_ID, PAD_TOKEN],
]);

// Tokens that should not be preceded by a space when detokenizing.
const NO_SPACE_BEFORE = new Set([...".,;:!?)]}'…", "''"]);
const NO_SPACE_AFTER = new Set([..."([{"]);

/**
 * The CANONICAL vocabulary serialization — byte-identical to the backend's
 * `GeoTokenizer.to_json()`.
 *
 * `vocab_sha256` is a digest of these exact bytes, so the spelling is part of the
 * format, not a formatting preference. This build used to emit `JSON.stringify`'s
 * output in its own key order while Python used `", "`/`": "` separators and sorted
 * keys, so the SAME model had two different digests depending on which build wrote the
 * file. Both stacks now emit: keys sorted, compact separators, and every non-ASCII
 * character escaped (Python's `ensure_ascii=True`) so the bytes never depend on the
 * encoding. The token regex admits any non-space symbol, so accented letters and
 * em-dashes really do reach the word list.
 */
export function canonicalVocabJson(words: readonly string[]): string {
  // Key order: "format" < "specials" < "words", and inside specials "<eos>" < "<pad>"
  // < "<unk>" — written out rather than sorted at runtime so the order is reviewable.
  const raw = JSON.stringify({
    format: "geo-tokenizer-v1",
    specials: { [EOS_TOKEN]: EOS_ID, [PAD_TOKEN]: PAD_ID, [UNK_TOKEN]: UNK_ID },
    words: [...words],
  });
  // JSON.stringify escapes control characters, quotes and backslashes exactly as
  // Python does; it leaves non-ASCII literal, which Python does not. Escape per
  // UTF-16 CODE UNIT (not code point), so an astral character becomes its two
  // surrogate escapes — which is exactly what `ensure_ascii` emits for one.
  //
  // The boundary is 0x7E, not 0x7F. Python's `ensure_ascii` is not "escape non-ASCII": its
  // encoder keeps `\x20`–`\x7e` and escapes everything else, DEL (U+007F) included. This
  // read `code < 0x80` and left DEL raw, so a word list containing it hashed differently in
  // the two stacks — and since a model's identity now covers its vocabulary, that split the
  // model id too: a file saved by the full stack was refused by the static build as "this
  // model file is corrupt", and vice versa. The geo token regex `[^\sa-z0-9]` admits any
  // single non-space symbol, so DEL really can reach the word list.
  let out = "";
  for (let i = 0; i < raw.length; i++) {
    const code = raw.charCodeAt(i);
    out += code < 0x7f ? raw[i] : "\\u" + code.toString(16).padStart(4, "0");
  }
  return out;
}

/** The one vocabulary format a model FILE may carry (`GeoTokenizer.to_json()`). */
export const VOCAB_FORMAT = "geo-tokenizer-v1";

/** Special-token ids a vocabulary may declare, under either spelling. */
const EXPECTED_SPECIALS: Record<string, number> = {
  [UNK_TOKEN]: UNK_ID,
  [EOS_TOKEN]: EOS_ID,
  [PAD_TOKEN]: PAD_ID,
  unk: UNK_ID,
  eos: EOS_ID,
  pad: PAD_ID,
};

/**
 * Refuse a vocabulary whose declared special ids are not the ones we use. Mirrors
 * `geo/tokenizer._validate_specials`: Python ignored the block entirely, so a file
 * declaring `"<unk>": 5` loaded there with 200 and was refused here — the two stacks
 * disagreed about whether a file was valid at all.
 */
function validateSpecials(specials: unknown, where: string): void {
  if (specials === undefined || specials === null) return;
  if (typeof specials !== "object" || Array.isArray(specials)) {
    throw invalidParam(`${where}: \`specials\` must be an object, got ${JSON.stringify(specials)}`);
  }
  const sp = specials as Record<string, unknown>;
  for (const [token, id] of Object.entries(EXPECTED_SPECIALS)) {
    if (sp[token] !== undefined && sp[token] !== id) {
      throw invalidParam(`${where}: special ${token} has id ${String(sp[token])}, expected ${id}`);
    }
  }
}

/** Deterministically split `text` into lowercase word/punctuation tokens. */
export function splitWords(text: string): string[] {
  const normalized = text.toLowerCase().replace(NORMALIZE_RE, (ch) => NORMALIZE_MAP[ch]);
  return normalized.match(TOKEN_RE) ?? [];
}

export interface EncodedText {
  ids: number[];
  texts: string[];
  unk: boolean[];
  n_unk: number;
  truncated: boolean;
}

/** Contract-shaped token list: [{id, text, unk}, ...]. */
export function encodedTokens(enc: EncodedText): { id: number; text: string; unk: boolean }[] {
  return enc.ids.map((id, i) => ({ id, text: enc.texts[i], unk: enc.unk[i] }));
}

export class GeoTokenizer {
  readonly words: string[];
  readonly idToText: Map<number, string>;
  readonly textToId: Map<string, number>;

  constructor(words: string[]) {
    if (words.length !== VOCAB_WORDS) {
      throw invalidParam(`GeoTokenizer requires exactly ${VOCAB_WORDS} words, got ${words.length}`);
    }
    if (new Set(words).size !== words.length) {
      throw invalidParam("GeoTokenizer vocabulary contains duplicates");
    }
    this.words = [...words];
    this.idToText = new Map(SPECIAL_TEXTS);
    this.words.forEach((word, offset) => {
      this.idToText.set(SPECIAL_TEXTS.size + offset, word);
    });
    this.textToId = new Map();
    for (const [tid, text] of this.idToText) {
      if (!SPECIAL_TEXTS.has(tid)) this.textToId.set(text, tid);
    }
    if (this.idToText.size !== VOCAB_SIZE) {
      throw invalidParam(`GeoTokenizer vocabulary resolves to ${this.idToText.size} ids, expected ${VOCAB_SIZE}`);
    }
  }

  /**
   * Defensive loader for the exported vocab.json. Accepts either export shape:
   *  - the backend tokenizer's `to_json()`: {"format": "geo-tokenizer-v1",
   *    "specials": {"<unk>": 0, ...}, "words": [1000 words]}, or
   *  - the static-site export: {"format": "geo-vocab-v1", "specials":
   *    {"unk": 0, ...}, "tokens": [all 1003 token texts, specials first]}.
   */
  static fromVocabJson(data: unknown): GeoTokenizer {
    if (typeof data === "string") data = JSON.parse(data);
    if (data === null || typeof data !== "object" || Array.isArray(data)) {
      throw invalidParam("vocab.json: expected a JSON object with a `words` or `tokens` array");
    }
    const obj = data as Record<string, unknown>;
    if (
      "format" in obj &&
      obj.format !== VOCAB_FORMAT &&
      obj.format !== "geo-vocab-v1"
    ) {
      throw invalidParam(`vocab.json: unknown tokenizer format ${JSON.stringify(obj.format)}`);
    }
    // Both spellings of the special map are validated when present — and a `specials`
    // that is not a map at all is refused rather than skipped.
    validateSpecials(obj.specials, "vocab.json");
    let words: unknown = obj.words;
    if (words === undefined && Array.isArray(obj.tokens)) {
      const tokens = obj.tokens;
      if (tokens.length !== VOCAB_SIZE) {
        throw invalidParam(`vocab.json: \`tokens\` has ${tokens.length} entries, expected ${VOCAB_SIZE}`);
      }
      const specialOrder = [UNK_TOKEN, EOS_TOKEN, PAD_TOKEN];
      specialOrder.forEach((tok, i) => {
        if (tokens[i] !== tok) {
          throw invalidParam(`vocab.json: tokens[${i}] is ${JSON.stringify(tokens[i])}, expected ${tok}`);
        }
      });
      words = tokens.slice(specialOrder.length);
    }
    if (!Array.isArray(words) || !words.every((w) => typeof w === "string")) {
      throw invalidParam("vocab.json: `words` (or `tokens`) must be an array of strings");
    }
    return new GeoTokenizer(words as string[]);
  }

  /**
   * The vocabulary inside a MODEL FILE — strict, and an exact mirror of the backend's
   * `GeoTokenizer.from_json`.
   *
   * `fromVocabJson` above is the loader for the shipped ASSET, and it deliberately
   * accepts the site's `geo-vocab-v1` / `tokens` export as well. Using it for a file the
   * user chose made the two stacks disagree about what a valid file is: a `tokens`-shaped
   * `vocab` block was accepted here (1000 words recovered) and was an untyped HTTP 500 in
   * Python. A file carries one format, and that format is `geo-tokenizer-v1`.
   */
  static fromModelVocabJson(payload: string): GeoTokenizer {
    if (typeof payload !== "string") {
      throw invalidParam(`a vocabulary must be JSON text, got ${typeof payload}`);
    }
    let data: unknown;
    try {
      data = JSON.parse(payload);
    } catch (e) {
      throw invalidParam(`vocabulary is not valid JSON: ${e instanceof Error ? e.message : String(e)}`);
    }
    if (data === null || typeof data !== "object" || Array.isArray(data)) {
      throw invalidParam(
        `vocabulary must be a JSON object with a \`words\` array, got ${Array.isArray(data) ? "array" : String(data === null ? "null" : typeof data)}`,
      );
    }
    const obj = data as Record<string, unknown>;
    if (obj.format !== VOCAB_FORMAT) {
      throw invalidParam(
        `Unknown tokenizer format ${JSON.stringify(obj.format)} (expected ${JSON.stringify(VOCAB_FORMAT)})`,
      );
    }
    validateSpecials(obj.specials, "vocabulary");
    const words = obj.words;
    if (words === undefined && "tokens" in obj) {
      throw invalidParam(
        "vocabulary has `tokens` but no `words`: `tokens` is the static site's asset shape " +
          `(specials included), while a model file carries the ${JSON.stringify(VOCAB_FORMAT)} ` +
          `\`words\` list of exactly ${VOCAB_WORDS} entries`,
      );
    }
    if (!Array.isArray(words) || !words.every((w) => typeof w === "string")) {
      throw invalidParam("vocabulary `words` must be an array of strings");
    }
    return new GeoTokenizer(words as string[]);
  }

  encode(
    text: string,
    opts: { truncate?: boolean; truncateSide?: "right" | "left"; maxTokens?: number } = {},
  ): EncodedText {
    const { truncate = true, truncateSide = "right", maxTokens = CONTEXT_WINDOW } = opts;
    let pieces = splitWords(text);
    let truncated = false;
    if (truncate && pieces.length > maxTokens) {
      truncated = true;
      pieces = truncateSide === "right" ? pieces.slice(0, maxTokens) : pieces.slice(-maxTokens);
    }
    const ids: number[] = [];
    const unk: boolean[] = [];
    let nUnk = 0;
    for (const piece of pieces) {
      const tid = this.textToId.get(piece);
      if (tid === undefined) {
        ids.push(UNK_ID);
        unk.push(true);
        nUnk++;
      } else {
        ids.push(tid);
        unk.push(false);
      }
    }
    return { ids, texts: pieces, unk, n_unk: nUnk, truncated };
  }

  /** Encode without truncation (fine-tuning token streams). */
  encodeStream(text: string): number[] {
    return this.encode(text, { truncate: false }).ids;
  }

  decode(ids: number[]): string {
    const out: string[] = [];
    for (const tid of ids) {
      const text = this.idToText.get(tid);
      if (text === undefined) {
        throw invalidParam(`Token id ${tid} is out of range (0..${VOCAB_SIZE - 1})`);
      }
      if (out.length === 0) out.push(text);
      else if (NO_SPACE_BEFORE.has(text) || NO_SPACE_AFTER.has(out[out.length - 1])) out.push(text);
      else out.push(" " + text);
    }
    return out.join("");
  }
}
